Add HTTP result header to file result types

boki opened the pull request:
The HTTP headers include interesting entries like X-Dropbox-Request-Id and
Dropbox-Api-Result for certain calls. This commit adds a Header field to the
output structs.

boki committed:
add HTTP result header to file result types

The HTTP headers include interesting entries like X-Dropbox-Request-Id and
Dropbox-Api-Result for certain calls. This commit adds a Header field to the
output structs.

tj commented:
hmm maybe we should normalize those header fields into individual struct fields, seems a little less ideal to just pass the http headers as-is IMO

boki commented:
I pondered that but Google's Drive API uses a similar approach for the headers (it includes the servers status code as well but with Dropbox this is not really necessary). A helper method to unmarshal the data from the `Dropbox-Api-Result` might be worth considering, though, as not every user might need that data.

tj commented:
I think that would be pretty useful, haven't needed it myself but I like the idea of having access to the download metadata etc from Dropbox-Api-Result instead of just a header

boki committed:
add UnmarshalDropboxApiResult helper function

The UnmarshalDropboxApiResult helper unmarshals the optional Dropbox-Api-Result
HTTP header in to the Metadata struct.

tj commented:
What about embedding the metadata struct into the regular output so we can unmarshal it for the user? I think that'd be a little better API-wise than having to manually unmarshal, then we can ditch exposing the Header

boki committed:
add X-Dropbox-Request-Id and Dropbox-Api-Result HTTP headers

boki committed:
add assert for APIResult to tests
